File: 6/6-6/Python/CongestionControl.py
import os

if __name__ == '__main__':
    MSS = 1024
    Threshold = 32768
    TriAckRound = 16
    TimeoutRound = 22
    EndRound = 26

    cwnd = 1
    ssthresh = int(Threshold / MSS)
    print("MSS: {}, Initial Threshold: {}".format(MSS, Threshold))
    print("For conveience, change unit from byte to segment, so initial threshold is {} segments".format(ssthresh))
    for i in range(1, EndRound + 1):
        print("Round: {}, cwnd: {}, send data".format(i, cwnd))
        if i == TriAckRound:
            ssthresh = int(cwnd / 2)
            cwnd = ssthresh
        elif i == TimeoutRound:
            ssthresh = int(cwnd / 2)
            cwnd = 1
        else:
            if cwnd < ssthresh:
                cwnd *= 2
            else:
                cwnd += 1
    os.system("pause")

File: 6/6-6/Java/CongestionControl.java
public class CongestionControl {
	private static int MSS = 1024;
	private static int Threshold = 32768;
	private static int TriAckRound = 16;
	private static int TimeoutRound = 22;
	private static int EndRound = 26;

	public static void main(String[] args) {
		int cwnd = 1;
		int ssthresh = Threshold / MSS;
		System.out.printf("MSS: %d, Initial Threshold: %d\n", MSS, Threshold);
		System.out.printf("For conveience, change unit from byte to segment, so initial threshold is %d segments\n",
				ssthresh);
		for (int i = 1; i <= EndRound; i++) {
			System.out.printf("Round: %d, cwnd: %d, send data\n", i, cwnd);
			if (i == TriAckRound) {
				ssthresh = cwnd / 2;
				cwnd = ssthresh;
			} else if (i == TimeoutRound) {
				ssthresh = cwnd / 2;
				cwnd = 1;
			} else {
				if (cwnd < ssthresh) {
					cwnd *= 2;
				} else {
					cwnd++;
				}
			}
		}
	}
}
